beacon-chain/core/electra/consolidations.go

```diff
@@ -0,0 +1,258 @@
+	"github.com/prysmaticlabs/prysm/v5/crypto/bls"
+	"github.com/prysmaticlabs/prysm/v5/math"
+	ethpb "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/time/slots"
+	"go.opencensus.io/trace"
+)
+
+// ProcessPendingConsolidations implements the spec definition below. This method makes mutating
+// calls to the beacon state.
+//
+// Spec definition:
+//
+//	def process_pending_consolidations(state: BeaconState) -> None:
+//	    next_pending_consolidation = 0
+//	    for pending_consolidation in state.pending_consolidations:
+//	        source_validator = state.validators[pending_consolidation.source_index]
+//	        if source_validator.slashed:
+//	            next_pending_consolidation += 1
+//	            continue
+//	        if source_validator.withdrawable_epoch > get_current_epoch(state):
+//	            break
+//
+//	        # Churn any target excess active balance of target and raise its max
+//	        switch_to_compounding_validator(state, pending_consolidation.target_index)
+//	        # Move active balance to target. Excess balance is withdrawable.
+//	        active_balance = get_active_balance(state, pending_consolidation.source_index)
+//	        decrease_balance(state, pending_consolidation.source_index, active_balance)
+//	        increase_balance(state, pending_consolidation.target_index, active_balance)
+//	        next_pending_consolidation += 1
+//
+//	    state.pending_consolidations = state.pending_consolidations[next_pending_consolidation:]
+func ProcessPendingConsolidations(ctx context.Context, st state.BeaconState) (state.BeaconState, error) {
+	ctx, span := trace.StartSpan(ctx, "electra.ProcessPendingConsolidations")
+	defer span.End()
+
+	if st == nil || st.IsNil() {
+		return nil, errors.New("nil state")
+	}
+
+	currentEpoch := slots.ToEpoch(st.Slot())
+
+	var nextPendingConsolidation uint64
+	pendingConsolidations, err := st.PendingConsolidations()
+	if err != nil {
+		return nil, err
+	}
+	for _, pc := range pendingConsolidations {
+		sourceValidator, err := st.ValidatorAtIndex(pc.SourceIndex)
+		if err != nil {
+			return nil, err
+		}
+		if sourceValidator.Slashed {
+			nextPendingConsolidation++
+			continue
+		}
+		if sourceValidator.WithdrawableEpoch > currentEpoch {
+			break
+		}
+
+		if err := SwitchToCompoundingValidator(ctx, st, pc.TargetIndex); err != nil {
+			return nil, err
+		}
+
+		activeBalance, err := st.ActiveBalanceAtIndex(pc.SourceIndex)
+		if err != nil {
+			return nil, err
+		}
+		if err := helpers.DecreaseBalance(st, pc.SourceIndex, activeBalance); err != nil {
+			return nil, err
+		}
+		if err := helpers.IncreaseBalance(st, pc.TargetIndex, activeBalance); err != nil {
+			return nil, err
+		}
+		nextPendingConsolidation++
+	}
+
+	if nextPendingConsolidation > 0 {
+		if err := st.SetPendingConsolidations(pendingConsolidations[nextPendingConsolidation:]); err != nil {
+			return nil, err
+		}
+	}
+
+	return st, nil
```

Comment:
This probably needs a broader design discussion and is outside of the scope of this PR, but what's the reason for returning the state here when the function doesn't make a copy of the passed in state? When I see a signature such as `func F(state) state`, I expect that I should only use the returned state because it will differ from the passed in state when the function returns. On the contrary, `SwitchToCompoundingValidator` doesn't return the state, so I expect the argument to be mutated inside the function.

It would be nice if we were consistent here.

P.S. The comment even says `This method makes mutating calls to the beacon state`

Reply:
There is even an argument that this method should be a receiver on the state.

```go
if err := st.ProcessPendingConsolidations(ctx, pendingConsolidations); err != nil {
  return fmt.Errorf("failed to process pending consolidations: %w", err)
}
```

I agree that this signature smells as is so I will change it as part of this PR such that it will not return the resulting state.

Reply:
One thing to note @rkapka, all of the other state transition functions return `(state.BeaconState, error)` so these methods will be the exception actually. It's not a big deal, we can also refactor the other methods not to return the state since it mutates it in place.